examples/Atlas/+atlasControllers/@InstantaneousQPController/InstantaneousQPController.m

```diff
@@ -0,0 +1,328 @@
+               'use_mex', @isscalar, ...
+               'solver', @(x) x == 0 || x == 1));
+      for f = fieldnames(options)'
+        obj.(f{1}) = options.(f{1});
+      end
+
+      obj.robot = r;
+      obj.param_sets = param_sets;
+      obj.robot_property_cache = atlasUtil.propertyCache(r);
+      import atlasControllers.*;
+      import atlasFrames.*;
+
+      if obj.debug
+        obj.debug_pub = ControllerDebugPublisher('CONTROLLER_DEBUG');
+      end
+
+      obj.controller_data = InstantaneousQPControllerData(struct('infocount', 0,...
+                                                     'qp_active_set', []));
+      obj.q_integrator_data = IntegratorData(r);
+      obj.vref_integrator_data = VRefIntegratorData(r);
+
+      obj.gurobi_options.outputflag = 0; % not verbose
+      if obj.solver==0
+        obj.gurobi_options.method = 2; % -1=automatic, 0=primal simplex, 1=dual simplex, 2=barrier
+      else
+        obj.gurobi_options.method = 0; % -1=automatic, 0=primal simplex, 1=dual simplex, 2=barrier
+      end
+      obj.gurobi_options.presolve = 0;
+      % obj.gurobi_options.prepasses = 1;
+
+      if obj.gurobi_options.method == 2
+        obj.gurobi_options.bariterlimit = 20; % iteration limit
+        obj.gurobi_options.barhomogeneous = 0; % 0 off, 1 on
+        obj.gurobi_options.barconvtol = 5e-4;
+      end
+
+      terrain = getTerrain(r);
+      obj.default_terrain_height = r.getTerrainHeight([0;0]);
+      if isa(terrain,'DRCTerrainMap')
+        terrain_map_ptr = terrain.map_handle.getPointerForMex();
+      else
+        terrain_map_ptr = 0;
+      end
+
+      obj.mex_ptr = ...
+             constructQPDataPointerMex(obj.robot.getMexModelPtr(),...
+                                       obj.param_sets,...
+                                       obj.robot_property_cache,...
+                                       obj.robot.getB(),...
+                                       obj.robot.umin,...
+                                       obj.robot.umax,...
+                                       terrain_map_ptr,...
+                                       obj.default_terrain_height,...
+                                       obj.solver==0,...
+                                       obj.gurobi_options);
+    end
+
+    function [y, v_ref] = updateAndOutput(obj, t, x, qp_input, foot_contact_sensor)
+      % Parse inputs from the robot and the planEval, set up the QP, solve it,
+      % and return the torques and feed-forward velocity.
+      % @param t time (s)
+      % @param x robot state vector
+      % @param qp_input a QPInput2D object
+      % @param foot_contact_sensor a 2x1 vector indicating whether contact force was
+      %                            detected by the [left; right] foot
+
+      % t0 = tic();
+
+      % Unpack variable names (just to make our code a bit easier to read)
+      r = obj.robot;
+      nq = obj.robot_property_cache.nq;
```

Comment:
Yeah, most of the time spent in the PlanEval loop was looking up body IDs and terrain contact points, which are all fixed at compile time. The cache gets that time down to basically zero